File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        //map node to parents for traversal till target;
        TreeNode* Root = root;
        vector<int> ans;
        unordered_map<TreeNode*,TreeNode*> mp;
        unordered_set<TreeNode*> visited;
        Parents(root,mp);
        Dfs(target,k,visited,mp,ans);
        return ans;
        
    }
    void Dfs(TreeNode* target,int k,unordered_set<TreeNode*>& visited, unordered_map<TreeNode*,TreeNode*>& mp,vector<int>& ans){
        // if node unvisited
        if(target == NULL) return;
        if(k==0 && visited.find(target)==visited.end()) ans.push_back(target->val);
        if(visited.find(target)!=visited.end()) return; // already visited so return;
        visited.insert(target);
        Dfs(target->left,k-1,visited,mp,ans);
        Dfs(target->right,k-1,visited,mp,ans);
        if(mp.find(target)!=mp.end()) Dfs(mp[target],k-1,visited,mp,ans);
        
    }
    void Parents(TreeNode* root, unordered_map<TreeNode*,TreeNode*>& mp)
    {
        if(root==NULL) return;
        if(root->left){
            mp[root->left]=root;
            Parents(root->left,mp);
        }
        if(root->right){
            mp[root->right]=root;
            Parents(root->right,mp);
        }
    }

};

File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        Set<TreeNode> vis = new HashSet<>();
        Map<TreeNode,TreeNode> mp = new HashMap<>();
        getParent(root,mp);
        List<Integer> ans = new ArrayList<>();
        Helper(target, k,mp,ans,vis);
        return ans;
     
    }
    void  Helper(TreeNode root, int target,Map<TreeNode,TreeNode> mp,List<Integer> ans,Set<TreeNode> vis){
        if(root==null) return ;
        if(target==0 && vis.contains(root)==false){
            ans.add(root.val);
            return;
        }
        if(vis.contains(root)==true) return;//already visited node
        vis.add(root);
        Helper(root.left,target-1,mp,ans,vis);
        Helper(root.right,target-1,mp,ans,vis);
         if(mp.containsKey(root)==true && vis.contains(mp.get(root))==false)
         {
            Helper(mp.get(root),target-1,mp,ans,vis);
         }
    }
    void getParent(TreeNode root, Map<TreeNode,TreeNode> mp)
    {
        if(root==null) return;
        if(root.left!=null){
            mp.put(root.left,root);
            getParent(root.left,mp);
        }
        if(root.right!=null){
            mp.put(root.right,root);
            getParent(root.right,mp);
        }
    }
}